haystack/preview/dataclasses/answer.py

```diff
@@ -0,0 +1,23 @@
+from typing import List, Optional, Union
+from dataclasses import dataclass
+from haystack.preview.dataclasses.document import Document
+
+
+@dataclass
+class ExtractiveAnswer:
+    answer: str
+    start: int
+    end: int
+
+
+@dataclass
+class GenerativeAnswer:
+    answer: str
+
+
+@dataclass
+class Answer:
+    answer: Optional[Union[ExtractiveAnswer, GenerativeAnswer]]
+    probability: float
+    question: str
+    documents: List[Document]
```

Comment:
The idea is nice but I think it's best done "in reverse": Let's have a base dataclass `Answer` that has all the shared fields, and specialize ExtractiveAnswer and GenerativeAnswer by the presence (or lack) of the position.

Something on the lines of:

```python
@dataclass
class Answer:
    probability: float
    question: str
    documents: List[Document]

@dataclass
class GenerativeAnswer:
    ...

@dataclass
class ExtractiveAnswer:
    position: Span
````

We may also specialize the two on document/documents (extractive should have a single doc, while generative can have many)

Reply:
How would you encode "no-answer" in this case?

Reply:
My reasoning (which I should probably have explained) was also that in practice when trying to differentiate between those case you would need to do some sort of exhaustive checking. This is a bit safer with this way of typing because type checkers will make sure it's exhaustive.
I don't really see an advantage of inheritance (except for less ugly syntax with the answer property) because there is no shared behaviour.

Reply:
> How would you encode "no-answer" in this case? Would it just be the base class?

Can't we just use `None`? Or am I missing something?

Reply:
Which field would be `None`? The base class does not have an answer field and we still need an answer object because "no-answer" still is a class with a probability.

Reply:
Why shouldn't the base class have data?

```
@dataclass
class Answer:
    data: Any
```

If `data` is `None` that's a no-answer. All other classes that inherit will use it as they see fit, they could even explicitly override the field type if necessary.

Reply:
Yes, we could do that. However, I am not sure if that's nicer than the composition approach. I am not sure if there might be cases where there is more than one obvious field which could be data (e.g., perhaps with tables) and where you would prefer being explicit instead of always calling it data.
It would also require making the dataclasses immutable because otherwise you would be restricting the parent's interface in a child class for the set operation.

Reply:
I think having a data field might actually easily devolve into the composition case. (If you create a new/use a more complex or non-standard class as the type of data.)

Reply:
> It would also require making the dataclasses immutable because otherwise you would be restricting the parent's interface in a child class for the set operation.

I don't see this as a problem to be fair. Also depends on what you mean with immutable. In any case I'd give the freedom to do that.

> I think having a data field might actually easily devolve into the composition case. (If you create a new/use a more complex or non-standard class as the type of data.)

That's up to us I think, it all depends on what we want to `Answer` class to be. If we just need it to for generative and extractive cases we could do without both inheritance and composition and have a single simple `Answer` dataclass like this.

```
@dataclass
class Answer:
    data: Optional[str]
    start: Optional[int]
    end: Optional[int]
    probability: float
    question: str
    documents: List[Document]
```

If instead we want to it to be a generic `Answer` that anyone can use in their `Component` to return a final answer I'd go with an extremely generic base `Answer` class and go from there.

Composition is nice but it's more limiting and also depends on how many other classes you have too. If you have few like in this case we can go with the above snippet.

Reply:
I think we'll probably have more answer cases. We already have TableQA in haystack which I think has its own return type and if our work on SQL generation goes as planned we'll also need to add an Answer type for this (might contain table or dataframe and potentially also original query).
So would you say we should then go with the `answer: Any` approach then?

Reply:
My two cents here: I think we're trying to unify two fairly different concepts, and that's why I like the idea of having `ExtractiveAnswer` and `GenerativeAnswer`. Even if they look very similar, the two objects have some fundamentally different semantics. For example:

1. ExtractiveAnswer can be a no-answer, but GenerativeAnswer not really (at least we don't see any situation so far when that would happen)

2. ExtractiveAnswer always comes from a single doc ans has a position where it was extracted from. GenerativeAnswer instead may be generated from several documents, but it might also come from no documents at all. It might come actually from some other random object/source, or from the LLM internal knowledge.

3. ExtractiveAnswers have well defined scores, while GenerativeAnswers doesn't.

and I'm sure there's more. Considering this, I think having a general Answer object for all of these wouldn't really be a good idea. I'd prefer to have a few common fields shared, if any at all, and have two Answer classes for these two concepts.

Reply:
> So would you say we should then go with the `answer: Any` approach then?

Yes, I'd go with that. I wouldn't call it `answer` though otherwise we're going to have an ugly and confusing `answer.answer` repetition way too often. `data` is much more sensible to me.

@ZanSara from the sound of it we don't even need `GenerativeAnswer`.

I expect `Answer` to be extremely simple then:

```
@dataclass
class Answer:
    data: Any
```

This can be used as a generative answer since it already has all it needs. To be fair we don't even need a class to represent a generative answer, it's a simple string. That is telling also by the fact that the `GenerativeAnswer` in the PR has a single field.

`ExtractiveAnswer` is necessary as it specialises on `Answer` by adding more information.

And this also will work in the future for `TableAnswer`, it could be something like this:

```
@dataclass
class TableAnswer(Answer):
    data: pd.DataFrame
    query: str
```

Having this hierarchy should also make it easier to handle connections between `Component`s.

Reply:
Wouldn't we still need a `GenerativeAnswer` class if `sources`/`documents` is separate? I only had one field in my PR for `GenerativeAnswer` because `documents` was already a field of `Answer`.

Reply:
Ah, certainly!